ballerina/rag.bal

```diff
@@ -0,0 +1,167 @@
+// in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+# Represents chunk retriever that finds relevant chunks based on query similarity.
+public type Retriever distinct isolated object {
+    # Retrieves relevant chunks for the given query.
+    #
+    # + query - The text query to search for
+    # + filters - Optional metadata filters to apply during retrieval
+    # + return - An array of matching chunks with similarity scores, or an `ai:Error` if retrieval fails
+    public isolated function retrieve(string query, MetadataFilters? filters = ()) returns QueryMatch[]|Error;
+};
+
+# Represents a retriever that finds relevant chunks based on query similarity.
+# The `Retriever` combines query embedding generation and vector search
+# to return matching chunks along with their similarity scores.
+public distinct isolated class VectorRetriever {
+    *Retriever;
+    private final VectorStore vectorStore;
+    private final EmbeddingProvider embeddingModel;
+
+    # Initializes a new `Retriever` instance.
+    #
+    # + vectorStore - The vector store to search in
+    # + embeddingModel - The embedding provider to use for generating query embeddings
+    public isolated function init(VectorStore vectorStore, EmbeddingProvider embeddingModel) {
+        self.vectorStore = vectorStore;
+        self.embeddingModel = embeddingModel;
+    }
+
+    # Retrieves relevant chunks for the given query.
+    #
+    # + query - The text query to search for
+    # + filters - Optional metadata filters to apply during retrieval
+    # + return - An array of matching chunks with similarity scores, or an `ai:Error` if retrieval fails
+    public isolated function retrieve(string query, MetadataFilters? filters = ()) returns QueryMatch[]|Error {
+        TextChunk queryChunk = {content: query, 'type: "text-chunk"};
+        Embedding queryEmbedding = check self.embeddingModel->embed(queryChunk);
+        VectorStoreQuery vectorStoreQuery = {
+            embedding: queryEmbedding,
+            filters: filters
+        };
+        VectorMatch[] matches = check self.vectorStore.query(vectorStoreQuery);
+        return from VectorMatch 'match in matches
+            select {chunk: 'match.chunk, similarityScore: 'match.similarityScore};
+    }
+}
+
+# Represents a knowledge base for managing chunk indexing and retrieval operations.
+public type KnowledgeBase distinct isolated object {
+    # Ingests a collection of chunks.
+    #
+    # + chunks - The array of chunk to index
+    # + return - An `ai:Error` if indexing fails; otherwise, `nil`
```

Comment:
We don't emphasize nil with backticks. We only emphasize constructs in the code like keywords, names (like variable names), etc.